src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,231 @@
+    * changed, was elected leader, lost leadership), and emits events accordingly.
+    *
+    * Materialized cancellable is used to abdicate leadership; which will do so followed by a closing of the stream.
+    */
+  def apply(
+    clientCloseable: LifeCycledCloseableLike[CuratorFramework],
+    zooKeeperLeaderPath: String,
+    zooKeeperConnectionTimeout: FiniteDuration,
+    hostPort: String,
+    singleThreadExecutor: ExecutorService): Source[LeadershipState, Cancellable] = {
+    val singleThreadEC = ExecutionContext.fromExecutor(singleThreadExecutor)
+    Source.queue[LeadershipState](16, OverflowStrategy.dropHead)
+      .mapMaterializedValue { sq =>
+        val emitterLogic = new CuratorEventEmitter(singleThreadExecutor, clientCloseable, zooKeeperLeaderPath, hostPort, sq)
+        emitterLogic.start()
+        sq.watchCompletion().onComplete { _ => emitterLogic.cancel() }(singleThreadEC)
+        emitterLogic
+      }
+      .initialTimeout(zooKeeperConnectionTimeout)
+      .concat(Source.single(LeadershipState.Standby(None)))
+      .via(EnrichedFlow.dedup(LeadershipState.Standby(None)))
+      .map { e =>
+        e match {
+          case LeadershipState.ElectedAsLeader =>
+            logger.info("Leader won.")
+          case LeadershipState.Standby(currentLeader) =>
+            logger.info(s"Leader defeated. Current leader: ${currentLeader.getOrElse("-")}")
+        }
+        e
+      }
+  }
+
+  @SuppressWarnings(Array("CatchThrowable"))
+  private class CuratorEventEmitter(
+      singleThreadExecutor: ExecutorService,
+      clientCloseable: LifeCycledCloseableLike[CuratorFramework],
+      zooKeeperLeaderPath: String,
+      hostPort: String,
+      sq: SourceQueueWithComplete[LeadershipState]) extends Cancellable {
+
+    val client = clientCloseable.closeable
+    private lazy val singleThreadEC: ExecutionContext = ExecutionContext.fromExecutor(singleThreadExecutor)
+    private lazy val leaderHostPortMetric: Timer = Metrics.timer(ServiceMetric, getClass, "current-leader-host-port")
+    private val curatorLeaderLatchPath = zooKeeperLeaderPath + "-curator"
+    private lazy val latch = new LeaderLatch(client, curatorLeaderLatchPath, hostPort)
+    @volatile private var isStarted = false
+    @volatile private var _isCancelled = false
+
+    /* It is important that we re-register our watch _BEFORE_ we get the current the current leader. In Zookeeper,
+     * watches are one-time use only.
+     *
+     * The timeline of events looks like this
+     *
+     * 1) We register a watch
+     * 2) We query the current leader
+     * 3) We receive an event (indicating child removal / addition)
+     * 4) Repeat step 1
+     *
+     * By re-registering the watch before querying the state, we will not miss out on the latest leader change.
+     */
+    def longPollLeaderChange(retries: Int = 0): Unit = Future {
+      try {
+        if (latch.getState == LeaderLatch.State.STARTED)
+          client.getChildren
```

Comment:
Do you think it's possible to refactor this using Promise into something like `def leaderEvent: Future[WatchedEvent]` ?

Reply:
We use `Future` as a helper to run in it in the serial execution thread devoted to curator concerns.

I'm not sure I see how a Promise would improve the readability. Can you provide an example? This is an asynchronous, trampoline-style recursive loop. I think that's a pretty good pattern.

Reply:
My idea was to move `new CuratorWatcher` part somewhere else to have something like this:
```scala
def watchLeaderEvent: Future[WatchedEvent] = {
  val p = Promise[WatchedEvent]()
  client.getChildren
  .usingWatcher(new CuratorWatcher {
    override def process(event: WatchedEvent): Unit = p.success(event)
  }
  .forPath(curatorLeaderLatchPath)
  p.future
}
```
And then we can keep the loop: 
```scala
def longPollLeaderChange(retries: Int = 0): Unit = Future {
      try {
        if (latch.getState == LeaderLatch.State.STARTED)
          watchLeaderEvent.foreach { _ => 
            if (!_isCancelled) longPollLeaderChange()
          }
        emitLeader()
      } catch {
        case ex: KeeperException.NoNodeException if retries < 100 =>
          // Wait for node to be created
          logger.info("retrying")
          Thread.sleep(retries * 10L)
          longPollLeaderChange(retries + 1)
        case ex: Throwable =>
          sq.fail(ex)
      }
    }(singleThreadEC)
```